File: example/cpp/src/product/play_mit_swing.cpp
#include <argparse/argparse.hpp>
#include <boost/asio.hpp>
#include <boost/json.hpp>
#include <chrono>
#include <cmath>
#include <csignal>
#include <iostream>
#include <thread>
#include <vector>

#include "airbot_hardware/executors/executor.hpp"
#include "airbot_hardware/handlers/arm.hpp"

using namespace std::chrono_literals;
using namespace airbot::hardware;

std::atomic<bool> running{true};

// Function to create command JSON
boost::json::object commandToJson(const std::string& command_type, const std::vector<double>& positions) {
  boost::json::object json_obj;

  // Convert position vector to JSON array
  boost::json::array pos_array;
  for (const auto& pos : positions) pos_array.push_back(pos);

  json_obj[command_type + "/position"] = pos_array;
  json_obj["timestamp"] = static_cast<double>(std::chrono::duration_cast<std::chrono::microseconds>(
                                                  std::chrono::steady_clock::now().time_since_epoch())
                                                  .count()) /
                          1e6;

  return json_obj;
}

bool is_arm_arrive(const std::array<double, 6> arm_target, const std::array<double, 6> arm_pose) {
  if (std::abs(arm_target[0] - arm_pose[0]) <= 0.02 && std::abs(arm_target[1] - arm_pose[1]) <= 0.02 &&
      std::abs(arm_target[2] - arm_pose[2]) <= 0.02 && std::abs(arm_target[3] - arm_pose[3]) <= 0.02 &&
      std::abs(arm_target[4] - arm_pose[4]) <= 0.02 && std::abs(arm_target[5] - arm_pose[5]) <= 0.02) {
    return 1;
  } else {
    return 0;
  }
}

void signalHandler(int signal) {
  if (signal == SIGINT) {
    running = false;
  }
}

int main(int argc, char* argv[]) {
  std::signal(SIGINT, signalHandler);
  argparse::ArgumentParser program("play_mit_swing", "1.0");

  program.add_argument("-i", "--can-interface")
      .help("CAN interface name (e.g., can0, can1)")
      .default_value(std::string("can0"));

  program.add_argument("-r", "--remote-ip").help("Plotjuggler IP address").default_value(std::string("127.0.0.1"));

  program.add_argument("-p", "--port").help("UDP port for Plotjuggler").scan<'i', int>().default_value(9870);

  program.add_argument("-T", "--period")
      .help("Period of sine wave oscillation (seconds)")
      .scan<'g', double>()
      .default_value(10.0);

  try {
    program.parse_args(argc, argv);
  } catch (const std::runtime_error& err) {
    std::cerr << err.what() << std::endl;
    std::cerr << program;
    return 1;
  }

  std::string can_interface = program.get<std::string>("--can-interface");
  std::string remote_ip = program.get<std::string>("--remote-ip");
  int port = program.get<int>("--port");
  double period = program.get<double>("--period");

  double ratio;
  if (period < 2) {
    std::cout << "Parameter T is set too small. The recommended value is between 2 and 20." << std::endl;
    return 0;
  } else if (period > 20) {
    std::cout << "Parameter T is set too big. The recommended value is between 2 and 20." << std::endl;
    return 0;
  } else {
    ratio = 2 * M_PI / period;
  }

  auto executor = airbot::hardware::AsioExecutor::create(8);
  auto arm = airbot::hardware::Arm<6>::create<MotorType::OD, MotorType::OD, MotorType::OD, MotorType::DM, MotorType::DM,
                                              MotorType::DM, EEFType::NA, MotorType::NA>();
  if (!arm->init(executor->get_io_context(), can_interface.c_str(), 250_hz)) {
    std::cerr << "Failed to initialize arm" << std::endl;
    return 1;
  }
  // Setup UDP socket
  boost::asio::io_context io_context;
  boost::asio::ip::udp::socket socket(io_context);
  boost::asio::ip::udp::endpoint server_endpoint(boost::asio::ip::make_address(remote_ip), port);
  try {
    socket.open(boost::asio::ip::udp::v4());
    std::cout << "UDP socket opened, sending data to " << remote_ip << ":" << port << std::endl;
  } catch (std::exception& e) {
    std::cerr << "Failed to open UDP socket: " << e.what() << std::endl;
    arm->disable();
    arm->uninit();
    return 1;
  }

  auto t = std::thread([&]() {
    while (running) {
      std::this_thread::sleep_for(std::chrono::milliseconds(10));

      auto state = arm->state();
      std::vector<double> positions(state.pos.begin(), state.pos.end());
      std::string state_string = boost::json::serialize(commandToJson("state", positions));

      try {
        socket.send_to(boost::asio::buffer(state_string), server_endpoint);
      } catch (std::exception& e) {
        std::cerr << "Failed to send state data: " << e.what() << std::endl;
      }
    }
  });

  arm->enable();
  arm->set_param("arm.control_mode", static_cast<uint32_t>(MotorControlMode::PVT));

  while (running) {
    if (!arm->state().is_valid) {
      arm->pvt({0.0, 0.0, 0.0, 0.0, 0.0, 0.0}, {M_PI / 4, M_PI / 4, M_PI / 4, M_PI / 4, M_PI / 4, M_PI / 4});
    } else {
      if (!is_arm_arrive({0.0, 0.0, 0.0, 0.0, 0.0, 0.0}, arm->state().pos)) {
        arm->pvt({0.0, 0.0, 0.0, 0.0, 0.0, 0.0}, {M_PI / 4, M_PI / 4, M_PI / 4, M_PI / 4, M_PI / 4, M_PI / 4});
      } else {
        break;
      }
    }
    std::this_thread::sleep_for(4ms);
  }

  auto synced = false;
  while (running) {
    auto t = std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::steady_clock::now().time_since_epoch())
                 .count();
    if (!synced && std::abs(arm->state().pos[0] - std::sin(static_cast<double>(t) / 1e3 * ratio)) > 0.01) {
      arm->set_param("arm.control_mode", static_cast<uint32_t>(MotorControlMode::PVT));
      arm->pvt({std::sin(static_cast<double>(t) / 1e3 * ratio), 0.0, 0.0, 0.0, 0.0, 0.0},
               {M_PI / 4, M_PI / 4, M_PI / 4, M_PI / 4, M_PI / 4, M_PI / 4}, {10.0, 10.0, 10.0, 10.0, 10.0, 10.0});
    } else {
      if (!synced) arm->set_param("arm.control_mode", static_cast<uint32_t>(MotorControlMode::MIT));
      synced = true;
      arm->mit({std::sin(static_cast<double>(t) / 1e3 * ratio), 0.0, 0.0, 0.0, 0.0, 0.0},
               {0.0, 0.0, 0.0, 0.0, 0.0, 0.0}, {0.0, 0.0, 0.0, 0.0, 0.0, 0.0}, {250.0, 250.0, 250.0, 50.0, 50.0, 50.0},
               {1.0, 1.0, 1.0, 1.0, 1.0, 1.0});
      std::vector<double> current_command = {std::sin(static_cast<double>(t) / 1e3 * ratio), 0.0, 0.0, 0.0, 0.0, 0.0};
      std::string command_string = boost::json::serialize(commandToJson("command", current_command));
      try {
        socket.send_to(boost::asio::buffer(command_string), server_endpoint);
      } catch (std::exception& e) {
        std::cerr << "Failed to send command data: " << e.what() << std::endl;
      }
    }

    std::this_thread::sleep_for(std::chrono::milliseconds(4));
  }

  // running = false;
  t.join();

  socket.close();
  arm->disable();
  arm->uninit();

  return 0;
}

File: example/python/product/play_mit_swing.py
import argparse
import json
import math
import socket
import threading
import airbot_hardware_py as ah
import time
import sys

# UDP server configuration (PlotJuggler's listening address)
UDP_SERVER_IP = "127.0.0.1"
UDP_SERVER_PORT = 9870

# Create a UDP socket
sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
server_address = (UDP_SERVER_IP, UDP_SERVER_PORT)

print(f"Sending data to {UDP_SERVER_IP}:{UDP_SERVER_PORT}")


def send_to_plotjuggler(arm, stop_event):
    while not stop_event.is_set():
        current_time = time.time()
        motor_num = 6
        state_positions = [0.0] * motor_num
        state_velocitys = [0.0] * motor_num
        state_forces = [0.0] * motor_num

        for i in range(motor_num):
            if arm.state().is_valid:
                state_positions[i] = arm.state().pos[i]
                state_velocitys[i] = arm.state().vel[i]
                state_forces[i] = arm.state().eff[i]

        # Create the JSON object
        json_obj = {
            "state/position": state_positions,
            "state/velocity": state_velocitys,
            "state/force": state_forces,
            "timestamp": current_time,
        }

        # Serialize the JSON object to a string and then encode to bytes
        message = json.dumps(json_obj).encode("utf-8")

        try:
            sock.sendto(message, server_address)
        except Exception as e:
            print(f"Failed to send state data: {e}")

        # Wait before the next round of prints
        time.sleep(0.01)


def main():
    # 设置命令行参数解析
    parser = argparse.ArgumentParser(
        description="Play 关节空间 MIT 控制 2 - 1 号关节正弦摆动"
    )
    parser.add_argument("-T", type=float, required=True, help="摆动的时间周期")
    parser.add_argument(
        "-i", "--can", default="can0", help="can口名称 (如: can0, can1)"
    )
    args = parser.parse_args()
    can_interface = args.can
    T = args.T
    if T < 2 or T > 20:
        print(
            "Parameter T settings are unreasonable. The recommended value is between 2 and 20."
        )
        sys.exit()

    executor = ah.create_asio_executor(1)
    io_context = executor.get_io_context()

    arm = ah.Play.create(
        ah.MotorType.OD,
        ah.MotorType.OD,
        ah.MotorType.OD,
        ah.MotorType.DM,
        ah.MotorType.DM,
        ah.MotorType.DM,
        ah.EEFType.NA,
        ah.MotorType.NA,
    )

    if not arm.init(io_context, can_interface, 250):
        raise RuntimeError("The arm initialization failed.")
    arm.enable()
    arm.set_param("arm.control_mode", ah.MotorControlMode.PVT)

    stop_event = threading.Event()
    print_thread = threading.Thread(target=send_to_plotjuggler, args=(arm, stop_event))
    print_thread.daemon = True
    print_thread.start()

    while not arm.state().is_valid or sum([abs(i) for i in arm.state().pos]) > 0.01:
        arm.pvt(
            [0.0, 0.0, 0.0, 0.0, 0.0, 0.0],
            [
                math.pi / 10,
                math.pi / 10,
                math.pi / 10,
                math.pi / 10,
                math.pi / 10,
                math.pi / 10,
            ],  # velocities
            [100, 100, 100, 100, 100, 100],
        )
        time.sleep(0.004)
    print("Arrived at zero position")

    arm.set_param("arm.control_mode", ah.MotorControlMode.MIT)

    try:
        now = time.time()
        start_time = now  # 记录循环开始的绝对时间
        while True:
            elapsed_time = time.time() - start_time  # 计算从循环开始已经过去的时间

            # 计算当前周期内的相对时间（0到10秒之间）
            time_in_period = elapsed_time % T
            # 计算角度：确保每个周期内完成一个完整的 2π 弧度
            angle = 2 * math.pi * (time_in_period / T)

            # 正弦值
            sin_value = math.sin(angle)

            arm.mit(
                [sin_value, 0, 0, 0, 0, 0],
                [0.0, 0.0, 0.0, 0.0, 0.0, 0.0],
                [0, 0, 0, 0, 0, 0],
                [250.0, 250.0, 250.0, 50.0, 50.0, 50.0],
                [1.0, 1.0, 1.0, 1.0, 1.0, 1.0],
            )
            time.sleep(0.004)
    except KeyboardInterrupt:
        print("\n检测到Ctrl+C，正在停止机械臂并释放资源...")
    finally:
        stop_event.set()
        if print_thread.is_alive():
            print_thread.join()

        arm.disable()
        arm.uninit()


if __name__ == "__main__":
    main()
